src/service/resources-impl.js

```diff
@@ -468,80 +483,105 @@ export class ResourcesImpl {
       this.ampdoc.getVisibilityState() != VisibilityState.PRERENDER ||
       resource.prerenderAllowed();
 
-    if (buildingEnabled && shouldBuildResource) {
+    const isCloseEnoughToViewport = this.buildWhenCloseToViewport_
+      ? ignoreQuota ||
+        resource.isBuildRenderBlocking() ||
+        resource.renderOutsideViewport() ||
+        (this.isIdle_() && resource.idleRenderOutsideViewport())
+      : true;
+
+    if (
+      buildingEnabled &&
+      !isBuiltOrBuilding &&
+      shouldBuildResource &&
```

Comment:
the name `shouldBuildResource` is confusing considering there are still 3+ other boolean variables that comprise whether or not a resource should be built.

in this case, `shouldBuildResource` really only refers to whether or not it should be excluded from prerender. Could we combine all these conditions into a a single function within `Resource` called something like `resource.shouldBuild()`?